### ISSUE TITLE
Pod 3 revision 3 issues

### ISSUE THREAD

**Issue by craigcabrey** [bobobo1618/ninesleep#2]
Hi,

Do you know which revision of the Pod 3 you have? I've pulled apart a Pod 3 but the daughterboard does not carry a microSD card. Do you have any pictures?

![20240127_163719](https://github.com/bobobo1618/ninesleep/assets/894565/db59edcb-7f8e-4709-b8be-75d0bee04267)

This is what I see. I'm wondering if they did a revision to remove the SD card.

**Comment by bobobo1618** [comment]
Damn, looks like they have a revision with a completely different daughterboard and it looks like they made this one themselves. Is there no SD card on it at all? Nothing on the back of the daughterboard for example?

I've added some info on my board to the readme to hopefully avoid anyone else running into the same problem.

Don't suppose you can share a photo of the back of the daughterboard as well? I'm curious what kind of processor they're using but the chip on this side is a Mediatek MT7668 which is just a wifi/bluetooth chip. If there's no processor on the other side, this board might just only act as a network adapter for the STM32 chips on the larger circuit board, so there wouldn't really be a way to hack it because there'd be nothing to hack.

The only workaround I can think of is to buy [the module used on my revision](https://www.variscite.com/product/system-on-module-som/cortex-a53-krait/var-som-mx8m-mini-nxp-i-mx8m-mini/), flash it and install it. Since you wouldn't be able to bootstrap it with the existing Eight Sleep hardware, you'd have to get [a dev kit](https://shop.variscite.com/product/evaluation-kit/var-som-mx8m-mini-evaluation-kits/) though and that's expensive and I can't guarantee that it'd work.

Something I hope to get to eventually is figuring out the interface the daughterboard uses to communicate with the rest of the hardware. If I can do that, it should theoretically be possible to make an open-source version of the daughterboard hardware and software.

**Comment by hongkongkiwi** [comment]
Yea, there may be a flash chip on there you can pull the firmware from. Be great to have a picture of the backside of that board..... @craigcabrey ?

Also, a non-blurry photo would be great!

**Comment by craigcabrey** [comment]
Sorry for the delay, here's the backside (hopefully not too blurry). I didn't see any kind of SPI chip to clip onto. It really seems like it's a fully custom daughterboard.

![20240401_175412](https://github.com/bobobo1618/ninesleep/assets/894565/5ef43473-b11f-4c0b-aa25-6c18ed43b5a5)

**Comment by craigcabrey** [comment]
That said, there are TX/RX pads, though I'm not sure if they are UART or something else. Maybe if somebody is in the ATX area we could collaborate lol

**Comment by andrew-kennedy** [comment]
@craigcabrey do you know what pod revision shows up in the app for you? I'd like to know if I have a compatible pod 3 without opening it up, if possible. My hub version is `20500-0003-F08-********` (redacted in case the full ID exposes anything private).

**Comment by nota-nota-nota** [comment]
@bobobo1618 is the SD card supposed to be ON the pop-up daughter board? Or under it?

For my board, the SD card is on the main board. In a slot that was hidden under the daughter board (no sd card on the daughter board).

![sdcard](https://github.com/bobobo1618/ninesleep/assets/111245470/5dca679d-3799-4691-820b-644ee43409c0)


My main control board is the same as same as the [original guide](https://github.com/bobobo1618/ninesleep#:~:text=Example%20board) "Control Board 230-0030 Rev-10.7-PRD"



@craigcabrey - There is an unpopulated SD card slot on your main board near the same location of the slot on my board. (it isn't exactly the same. Your SD slot is closer to the screw than mine is). 
Unlikely, but connecting an SD card might work... 

What revision of the main board do you have?

![image](https://github.com/bobobo1618/ninesleep/assets/111245470/2338f24e-04b4-4845-88f0-5f9c0fd818cd)

**Comment by misslisa62tx** [comment]
I'm fairly certain this is rev 3 of the control board, same version I have, where yours is version 2.  (white sticker in the upper left)
![Control Board Hi Res](https://github.com/user-attachments/assets/47e3f849-5148-4622-9165-ed18c659951d)

Turs out there are a bunch of missing components for the SD card, not just the slot. (at least on mine, not sure about his)
![Missing SD Components](https://github.com/user-attachments/assets/ac341102-22d7-4c85-98a2-f4f3334a7962)

What he didn't show clearly on the daughter board is the Kingston 32GB EMMC flash chip.
![Kingston 32GB EMMC](https://github.com/user-attachments/assets/574b80f3-9662-4ebc-9514-e3e9f0500651)

I'm fairly certain that's where the code will be now.  I don't see any other memory chips on either board.
Unfortunately, I believe it would have to be desoldered in order to read/write it.

**Comment by misslisa62tx** [comment]
Oh, I meant to say.  They've added another set of jumper pads on the main control board in J7.  My guess is that is now the factory programming port.  I might trace that out and see if I can prove that.

**Comment by timwu** [comment]
Out of curiosity, can we get the manufacturing dates of the different Pod 3s in question? Might help to discern which Pod 3's are usable before tearing them open.

**Comment by misslisa62tx** [comment]
I now have two units.  One dated 20220920 which is the version with the SD Card.  The other dated 20240307 which does not have an SD Card.

My 2022 unit was sold cheap as faulty (I haven't booted it, so I don't know for sure yet).  If it boots far enough to flash the processor card and they are compatible, I'm planning to try to program the 2024 unit processor card with the 2022 unit control board or run the 2022 unit processor on the 2024 unit control board.  

Right now I've got other things on my plate that have priority.  I'll update when I get the chance to try that out.

**Comment by timwu** [comment]
Thanks! I was meaning to report back on my question above. Sorry if this has diverged a bit off topic.

TL;DR: Pod 3's with FCC ID: TFB-1004 on the back seem to be the moddable ones.

After looking through some listings for used pod 3 hubs, I noticed that there are some (like yours, manufactured ~2022) that have an FCC ID on the back sticker of TFB-1004 like this:
![pod3back](https://github.com/user-attachments/assets/3804f709-490e-423c-b341-e1bf48715a4f)

This is in contrast to more recently manufactured ones with an FCC ID of `2AYXT61100001`, like this:
![s-l1600 copy](https://github.com/user-attachments/assets/153adcd0-8eca-4921-8400-30354be6bfc7)

Looking up FCC ID: TFB-1004 leads to [this](https://fcc.report/FCC-ID/TFB-1004). Which looks like the regulatory filing for the Variscite MXM board. Going off that, I bought a used one with the corresponding FCC ID and can confirm it does indeed have the SD card slot and such.

Also interesting, looking up FCC ID `2AYXT61100001`, leads to [this](https://fccid.io/2AYXT61100001), which has some pictures of the eight sleep board. Not sure if there's anything else useful in that filing though.

**Comment by schrauger** [comment]
> I now have two units. One dated 20220920 which is the version with the SD Card. The other dated 20240307 which does not have an SD Card.
> 
> My 2022 unit was sold cheap as faulty (I haven't booted it, so I don't know for sure yet). If it boots far enough to flash the processor card and they are compatible, I'm planning to try to program the 2024 unit processor card with the 2022 unit control board or run the 2022 unit processor on the 2024 unit control board.
> 
> Right now I've got other things on my plate that have priority. I'll update when I get the chance to try that out.

I'd be very interested to know if this works. I'm looking for pods with the older FCC ID showing to purchase, but if we can use the older board on a newer pod, it would mean any future repairs or replacements could be an easy swap out, without worrying about finding an ever-decreasing supply of older revision pods.

**Comment by timwu** [comment]
I guess another option would be to see if the control board + processor card can be swapped. Maybe the interface for the processor card is proprietary/subtly incompatible, but the lower level connectors (i.e. pump on, fan on, etc.) could be compatible between the older control board and newer pod unit? When I get around to it I can try to take out my control board and compare the connectors with the ones for the newer control board in the FCC filing.

I'd imagine the breakdowns we would expect in the longer term would be the larger mechanical components anyways. Maybe that would prove sufficient.

**Comment by morelazers** [comment]
Also very interested in this. I have a newer `2AYX...` model Pod 3 and after cancelling the subscription lost access to basically all of the features. Happy to perform minor surgery on my unit though I'm useless with a soldering iron.

**Comment by davidsilva2841** [comment]
Hi @bobobo1618,

I was wondering if you could share a complete copy of the SD card in your Eight Sleep.

I’m trying to hack into my own Eight Sleep, but I have the revision E board, which does not have an SD card. I’m currently attempting to gain access to the OS via the J7 and J10 interfaces.

Having a complete copy of the SD card might provide documentation or clues on how to connect to and communicate with the device.

I won't share or publish the copy anywhere. I’m just looking for guidance since hardware hacking is new to me.

Thank you!

**Comment by timwu** [comment]
@davidsilva2841 Check out [this blogpost](https://blopker.com/writing/04-zerosleep-1/), there's a bit of info on how to get the current swupdate image among other interesting tidbits.

**Comment by davidsilva2841** [comment]
@timwu 
The file it links to [here](https://blopker.com/writing/04-zerosleep-1/#:~:text=what%E2%80%99s%20there%20yourself%2C-,grab%20it%20here,-.%20Use%20cat%20firmware), does not exist anymore.

Would you be able to provide a copy of the SD card?

**Comment by smiba** [comment]
The file still is there just add some numbers to the deviceid :)

It seems they just blocked "1"

**Comment by davidsilva2841** [comment]
I'm connected to the J7 JTAG using a Tag-Connect TC2070-IDC with my FT2232H-56Q and am trying to determine the pin mapping. I plan to try the J10 JTAG later.

**If anyone with access to their SD card could share copies of the /proc/, /sys/, and /dev/ folders, it would be greatly appreciated, as these might contain documentation or hints about the pins. These folders were not included when I downloaded the firmware update following [this](https://blopker.com/writing/04-zerosleep-1/#:~:text=what%E2%80%99s%20there%20yourself%2C-,grab%20it%20here,-.%20Use%20cat%20firmware)**

Working with JTAGs is new to me, so any help or guidance would be greatly appreciated.

**Comment by davidsilva2841** [comment]
I've identified a SWD interface on the J10 interface using an arduino and [this script](https://github.com/szymonh/SWDscan)

```
| CLK PIN | IO PIN | ACK | PART NO | MAN ID |
|    4    |    8   |  1  |   bb11  |   23b  | <- Success
```

 It shows up consistently when you provide 3.3V power to pad 10 & the device is powered off.

You can easily connect to the J10 interface with [this Tag-Connect cable](https://www.tag-connect.com/product/tc2050-idc-tag-connect-2050-idc). No soldering required. 

Working on trying to communicate on that interface now. Feel free to reach out if you'd like to help

**Comment by bobobo1618** [comment]
@davidsilva2841 you mean J10 on the mainboard (the one with "CONTROL BOARD" on the silkscreen), not the daughterboard in what looks like a RAM DIMM slot?

I believe that J10 will connect you to one of the STM32s that interfaces directly with the hardware. The daughter board doesn't talk directly to the various hardware components but rather talks to at least one STM32 over UART. IIRC there are two UART interfaces, one with a name indicating that it's mainly for sensors, the other with a name indicating that it's actually driving pumps and heating and whatnot. I'm not sure if those are different MCUs or just two interfaces on one.

Just be aware that I don't think J10 gives you a way to communicate with the board that runs Linux. You'll be able to play around with the STM32s that interface directly with the hardware but I suspect there's a lot of actual work going on in there and it's much harder to work with. You won't get a nice terminal like I did with the daughter board. You'll have a debugger and you'll probably be able to dump the firmware but I don't think there's a whole lot you can do with that unless you do some other hardware modifications.

Honestly, it might be quicker to reverse engineer the MCU by observing how it communicates with the rest of the board using an oscilloscope and/or logic analyzer, then reimplementing that with your own code and just overwriting what Eight Sleep put on the MCU.

The most helpful thing to play with is the daughter board though. If you can figure out the model of the specific board they're using, you might be able to source a devkit or something.

**Comment by davidsilva2841** [comment]
J10 on the main board, not the daughterboard.

> You won't get a nice terminal like I did with the daughter board.

Are you talking about how you got a terminal session by modifying the OS with the SD card on your model?

Do you think J7 could connect to the daughterboard?

**Comment by davidsilva2841** [comment]
*It also turns on a light on the daughterboard when you power it via pad 10 on J10. The device is still unplugged though.  Is it still unlikely that there's any sort of interface on the daughterboard?
![IMG_2186](https://github.com/user-attachments/assets/1b34b0f5-ac02-4d45-8184-05bec2fde00c)

**Comment by bobobo1618** [comment]
J10 is present on earlier revisions, is located right next to the STM32 and has no obvious signs of traces heading towards the daughterboard so I feel fairly confident that it's for the STM32 and not the daughterboard. 

The light on the daughterboard may just be due to power being supplied?

J7 does seem much more likely a candidate. It's not present on earlier board revisions and I think I do see traces heading towards the daughterboard. No idea what the pinout is though.

**Comment by davidsilva2841** [comment]
If I shared the voltage measurements & resistance measurements for J14, would you be interested in taking a look?

**Comment by bobobo1618** [comment]
Voltage and resistance wouldn't help me at all I'm afraid.

I think the most helpful starting point would be clear, high-res photos of both sides of the daughterboard. It doesn't need to be perfect but if you can read the part numbers, you can find data sheets for the parts. If you can find the data sheets, you can get the pinouts. Pinouts + a multimeter with a fine probe could help figure out if useful signals make it out to the main board somewhere (e.g. if you get continuity from the pin that the data sheet says is SWD to one of the pads on J7, you know J7 has a connection for a debugger on it).

Those test pads on the daughterboard could also be useful. I think I see a pair for USB and I'd be surprised if UART wasn't there somewhere.

UART could be really useful but I'm guessing we'd need to figure out what the password is. I can probably get the hash easily enough but I don't think I have enough computing power to brute force it.

I actually seem to have lost SSH access to my own pod... I hard-coded a public key to which I no longer have the private key.

**Comment by davidsilva2841** [comment]
![control_board_with_daughter](https://github.com/user-attachments/assets/a87d284c-11ac-4802-8c3a-85c742f809aa)
![control_board_no_daughterboard](https://github.com/user-attachments/assets/4181c2a2-df97-4a17-87d5-3acde00d3995)

**Comment by davidsilva2841** [comment]
![daughterboard_top](https://github.com/user-attachments/assets/bff5fa10-9e67-41e3-8a42-93268b978ac8)
![daughterboard_bottom](https://github.com/user-attachments/assets/e78cbc98-3569-472e-9cec-1fbaf0e4faf2)

**Comment by davidsilva2841** [comment]
![image](https://github.com/user-attachments/assets/168af2c0-368c-439f-9b92-2f02b274932e)

https://fccid.io/2AYXT61100001/Internal-Photos/Internal-Photos-6388828
It looks like the board they pictured in their FCC filing has some sort of micro SD card slot soldered onto the board too. My board has a slot for it but there's nothing there. Maybe we could solder a micro SD card slot to the board and try to flash a modified version of the OS?

**Comment by davidsilva2841** [comment]
Continuity results so far (J7 interface). I'll try to figure out how to connect to the JTAG or UART using this 
```
touch_pad , label     , jtag_pad
TP16      , DBGTX     ,        8
TP3       , CBGRX     ,        6
TP39      , USBD+     ,       12
TP40      , USBD-     ,       14
TP34      , RST       ,        2
TP38      , PGM       ,        4
```

**Comment by davidsilva2841** [comment]
I was able to capture around 2K lines of readable UART output using a baud rate of 115200. After that, the UART output becomes garbled, displaying invalid characters. I also experimented with other standard baud rates (e.g., 9600, 19200), but the issue persists.

While exploring the OS, I discovered a script that seems to disable the UART: `/lib/systemd/system/disable-serialgetty.sh`. I suspect this might be related to the issue, as it could be stopping or reconfiguring the UART service during boot. I'm also not sure I have a correct version of the OS for this model. 

I've tried spamming interrupt keys during the boot process up to the point where the output becomes garbled, but I still haven't been able to access a shell.

I've attached a copy of the readable portion of the UART output for reference. If anyone has any ideas I'm happy to try them.

[uart_output.txt](https://github.com/user-attachments/files/18240842/uart_output.txt)

**Comment by davidsilva2841** [comment]
The intial baud rate is 115200, then it switches to 921600. Looks like UART is behind a login... I'm going to see if I can interrupt the boot or something

```

                           (                          (
                   )     ) )\ )  (                    )\ )       (        )
 (   (   (  (   ( /(  ( /((()/(  )\   (    (         (()/(       )\ )  ( /(
 )\  )\  )\))(  )\()) )\())/(_))((_) ))\  ))\ `  )    /(_)) (   (()/(  )\())
((_)((_)((_))\ ((_)\ (_))/(_))   _  /((_)/((_)/(/(   (_))   )\   ((_))((_)\
| __|(_) (()(_)| |(_)| |_ / __| | |(_)) (_)) ((_)_\  | _ \ ((_)  _| ||__ (_)
| _| | |/ _` | | ' \ |  _|\__ \ | |/ -_)/ -_)| '_ \) |  _// _ \/ _` | |_ \
|___||_|\__, | |_||_| \__||___/ |_|\___|\___|| .__/  |_|  \___/\__,_||___/
        |___/                                |_|
             Version: 6ba61c8e8b3295d6210fead782dd122af55d52a5
eight-pod login:
```

*Update: You can easily interrupt the boot process and get a U-boot shell

**Comment by nota-nota-nota** [comment]
Hi David and Bobo, thanks for sharing your progress so far. 

I just wanted to share that I have torn down a pod 4. it uses the same HW revision for the daughterboard.
"i350 SOM MODULE". (Model 830-00065-00), HW: E02

The SW is updated (Likely for Pod 4 SW/HW). My board had an old sticker with "SW: 6ba61cB" (which matches the Version: shown on your board and in the UART output). There was a new sticker placed over this one. Potentially 8 sleep re-flashed this board to be used in a pod 4 rather than a 3? "SW: 

The main board is completely different, but that was expected. I can see J7  appears to be the same between the Pod3 RevE and the Pod4 (14 pin jtag connector)

I'll see if I can get UART on J7. @davidsilva2841 can you share how you "[interrupted the] boot process and get a U-boot shell" ?


![image](https://github.com/user-attachments/assets/b379461c-8baf-400f-86f1-8db434eaf295)

<img width="811" alt="image" src="https://github.com/user-attachments/assets/f92c22b5-ce3a-473b-aaf2-d2edbc4d4bab" />


![image](https://github.com/user-attachments/assets/45f023c0-4e62-46e3-a13d-5da5fdb10c3e)

**Comment by davidsilva2841** [comment]
# How to get root

1. Connect to J7 header on board with [TC2070-IDC](https://www.tag-connect.com/product/tc2070-idc)

3.  Use pins 1, 6, 8 for UART access (1 => GND, 6 => RX, 8 => TX). Baud rate is 921600. I used this [$13 FTDI FT232RL](https://www.amazon.com/gp/product/B07TXVRQ7V/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) for connecting to the UART. (FYI - you need to connect the TX pin from the 8 sleep to the RX pin on the FT232RL, and the RX pin connects to the TX pin)
![control_board_with_daughter](https://github.com/user-attachments/assets/a6fb867f-7d4f-4fcb-b642-896b78c29c2a)

4. Interrupt the boot process to get a U-boot shell.  I used `minicom` on my mac for this. Heads up: I tried the `screen` tool on my mac instead of `minicom` and it did not work correctly. 
```
# Power on your 8 sleep pod AFTER you run this command:
$ minicom -b 921600 -o -D /dev/tty.usbserial-B0010NHK
...
...
# Hit CTRL+C when you see this:
Hit any key to stop autoboot:  0
```

5. Once in the U-boot shell, modify the boot args to open up a root shell. This allows us to bypass the login
```
=> setenv bootargs "root=PARTLABEL=rootfs_a rootwait init=/bin/bash"
=> run bootcmd
```

6. In your shell session, remount the file system as read-write
```
# Mount /proc for process and system information
mount -t proc proc /proc

# Mount /sys for hardware and system-level information
mount -t sysfs sysfs /sys

# Mount /dev for device files (e.g., /dev/mmcblk0)
mount -t devtmpfs devtmpfs /dev

# Mount /run (optional but useful for some runtime scripts)
mount -t tmpfs tmpfs /run

# Mount the file system in read write mode (BE CAREFUL EDITING MODIFYING FILES)
mount -o remount,rw /
```


7. Change the password via cli
```
passwd root
passwd rewt
```

8. Reboot
```
sync
reboot -f
```

9. At the 8 sleep login prompt, login as root with the new password you set
```
eight-pod login: root
Password:
root@eight-pod:~#
```

10. Disable firmware updates
```
systemctl disable --now swupdate-progress swupdate defibrillator eight-kernel telegraf vector frankenfirmware
systemctl mask swupdate-progress swupdate defibrillator eight-kernel telegraf vector frankenfirmware
```

11. Connect to the internet
```
nmcli dev wifi list
nmcli dev wifi connect "<WIFI_SSID>" password "<PASSWORD>"
nmcli connection modify <WIFI_SSID> connection.autoconnect yes

# Test your connection 
ping -c 5 8.8.8.8
```

**Comment by davidsilva2841** [comment]
@nota-nota-nota You may want to double check the UART pinout on your board matches mine. You can do that with a multi meter test for continuity between the J7 pin 6 (RX) and TP3 / DBGRX pad on your daughterboard. Same for J7 pin 8 (TX)

**Comment by nota-nota-nota** [comment]
> @nota-nota-nota You may want to double check the UART pinout on your board matches mine. You can do that with a multi meter test for continuity between the J7 pin 6 (RX) and TP3 / DBGRX pad on your daughterboard. Same for J7 pin 8 (TX)

I just checked. I do have continuity between the test pads (GND, TX, and RX). Thank you. I'll connect the pins to a FTDI this weekend and see if I can reproduce your step 1-6. 

* edit: Updated to include the check for ground too. The layout appears to be the same, just rotated 90 degrees clockwise.

![image](https://github.com/user-attachments/assets/c4b373fd-3200-4942-97c4-c9a61963757a)

**Comment by davidsilva2841** [comment]
*Edit: Might actually be the same, it’s just rotated…. I need sleep…..

Looks like you have a different pinout. You’ll need to identify your ground too. IIRC, you can check that by finding a pin that shows 0 when you measure continuity to one of your known ground test points. Looks like there’s a ground point on the main board in the top left and a ground test point on the daughter board near the bottom left (TP30)

You may have a different baud rate too, and you’d need a logic analyzer to identify the baud rate. I used this with PulseView on my mac
https://www.amazon.com/dp/B077LSG5P2?ref=ppx_pop_mob_ap_share

**Comment by bobobo1618** [comment]
FWIW, here's `rewt` from `/etc/shadow`:

```
rewt:$6$8zHn61PsWU3gtHoY$1zrMb/9vnvx6uR7gFDNtimIOstoqyxCrLgLYuq9mJx2jcqPXkPGlueyVx8V8QcDMfVaD7vPZCHGCl5TtUsrCk/:19747:0:99999:7:::
```

Not sure if every board will have the same password but if someone has some hefty hardware to brute-force this, we can find out.

**Comment by davidsilva2841** [comment]
My `rewt` in `/etc/shadow` is different than that
`rewt:$1$XshBnCRI$78uAK/0o.UED0M.fIO7rg1:15069:0:99999:7:::`

**Comment by nota-nota-nota** [comment]
Updated my [photo above](https://github.com/bobobo1618/ninesleep/issues/2#issuecomment-2564251389). Pod 4 main board does have the same JTAG pinout on J7 (rotated 90 degrees clockwise).

**Comment by bobobo1618** [comment]
> My `rewt` in `/etc/shadow` is different than that `rewt:$1$XshBnCRI$78uAK/0o.UED0M.fIO7rg1:15069:0:99999:7:::`

Hmm, I may have tried to update mine to something else. Yours is the better one to go off.

Plus it looks like yours is just MD5, so much easier to break.

**Comment by davidsilva2841** [comment]
The password might not need to be cracked. I've confirmed that I can modify the file system and connect to the internet from the shell session I accessed by interrupting U-Boot.

The steps for the internet were specific to my device version, I think we have different kernel modules. I'll update the steps from my earlier comment.

I'm going to backup the entire storage before I try modifying anything else. I've only modified the `/etc/wpa_supplicant.conf` so far

**Comment by davidsilva2841** [comment]
*Updated my instructions earlier. Added getting root access & connecting to the internet

**Comment by davidsilva2841** [comment]
@bobobo1618 How is the API accessed here?
https://github.com/bobobo1618/ninesleep?tab=readme-ov-file#api

I can ssh into my device. I tried to GET `192.168.50.60/variables` and am getting `Error: connect ECONNREFUSED 1`

Is that how the API is supposed to be used? Or is there something in the configs or firewall I need to change first?

**Comment by throwaway31265** [comment]
Excellent work everyone! 

I've been able to get root access to my pod 3 and have been building a local express.js server that serves a react app for controlling the pod. The server runs on the 8 sleep device itself.

The app **does not** depend on internet access and can be used on **LAN**. This app allows for **100% control** of the device.  So far, I've created a solid personal react app for controlling the device I've created for **research purposes** and integrations with **third party software**. So your phone or any device on connected to your WiFi could access the app at the pods LAN IP 192.168.1.XX and control the device from the web app. The app was built mobile first but also works on desktop.

I'm researching how schedules and alarms could be created next and am looking forward to sharing my research

**Comment by throwaway31265** [comment]
Hi everyone, here's the app I made so far. I tested this on a the pod 3 without the SD card (FCC ID: 2AYXT61100001)
https://github.com/throwaway31265/free-sleep

I followed @davidsilva2841 comment [here](https://github.com/bobobo1618/ninesleep/issues/2#issuecomment-2564228235)


Some features so far (none of which require the 8 sleep subscription OR INTERNET connection)
- Allows complete control of device WITHOUT requiring internet access. If you lose internet, your pod WILL NOT turn off, it will continue working! You can completely block WAN internet access if you'd like too. (I blocked all internet access from my pod on my router...)
- WARNING: This will bypass blocked devices, please use responsibly.
- Dynamic temperature control with real-time updates.
- Schedule management: Set power on/off times and temperature adjustments.
- Settings customization: Configure timezones, away mode, and priming schedules.
- Responsive design for desktop and mobile.

*Added installation instructions here
https://github.com/throwaway31265/free-sleep/blob/main/INSTALLATION.md

Added images here for connecting to the jtag interface 
https://github.com/throwaway31265/free-sleep/tree/main/docs/pod_teardown

Thanks @bobobo1618 for sharing all your findings around the API

I've been using the app for the past week as I've been developing it. I have more updates planned. Feel free to create pull requests or bug tickets. Thanks!

**Comment by throwaway31265** [comment]
@nota-nota-nota If you need any help with the pod 4, feel free to reach out. I'm happy to take a look or coordinate on discord

**Comment by throwaway31265** [comment]
Confirmed this is working on pod 4 with @nota-nota-nota  earlier, one extra step for pod 4 has been added to the installation instructions here
https://github.com/throwaway31265/free-sleep/tree/main

(This app is very new, I would hold off on being an early adopter unless you have programming experience or are comfortable with debugging. Everything should be easily reversible by resetting the firmware though)

**Comment by RevoGap** [comment]
![70BF4576-0A0B-4D24-BA95-7F5AAF1CF05F](https://github.com/user-attachments/assets/fc4ff9b2-0c3d-4a48-ae4e-bf0bd138bdda)
![FC9A5C8F-AB76-4E52-9FA3-39F5AF964477](https://github.com/user-attachments/assets/bd296457-d402-4f3b-a4d4-7259d8cc5ee1)

In response to your wonderful updates, I disassembled my Pod 4 Hub. This was way easier than the Pod 3, only having to take off the front mesh panel. There are no metal clips anywhere, so for anyone trying the same:

Start from one side with some plastic wedges and work around the front to the other side: it will pop off easily! There are 4 +-2cm wide clips spaced evenly on both sides and on the curved top and bottom edges respectively. I did not damage any of these by carefully working from one clip to the next. 

My board and daughter board are similar to nota-nota-nota’s, but I have uploaded them here for reference. Tomorrow, the TC2070-IDC will come in, so I will try to get root then (as a total newbie to any of this). Will update you on my progress here! Thank you for your hard work.

**Comment by throwaway31265** [comment]
Nice! Feel free to reach out if you need help

**Comment by throwaway31265** [comment]
@RevoGap Is there a small button in the back of the pod 4 to firmware reset the pod?

**Comment by RevoGap** [comment]
> @RevoGap Is there a small button in the back of the pod 4 to firmware reset the pod?

There is!

![IMG_2026](https://github.com/user-attachments/assets/e8140e61-8ab1-4f1f-a069-6a02800da5de)

**Comment by RevoGap** [comment]
Working on this now. Will need some help. Better to do this through DM? In any case:
Connected the pins as discussed in steps 1&2
Installed minicom for Mac (Silicon, M1)
In step 3, you mention command
$ minicom -b 921600 -o -D /dev/tty.usbserial-B0010NHK
But using ls /dev, I found my device to show up as tty.usbserial-B0000BVQ. So I edited the command accordingly.
This brings me to the following screen:

Welcome to minicom 2.9

OPTIONS: 
Compiled on Sep 22 2023, 21:10:41.
Port /dev/tty.usbserial-B0000BVQ, 12:50:00

Press Meta-Z for help on special keys

Every keystroke I make in the terminal window now lights up the LED on the TC2070 for the TXD, but nothing shows up in the terminal window. After running the command and powering on the Hub 4, no message shows up saying "Hit any key to stop autoboot:  0"

So this is where I am currently stuck at.

**Comment by morelazers** [comment]
@RevoGap I just succeeded with a Pod 3, I'm happy to DM and try to help you debug on Discord: xcessiveoverlord (or 238452185900253185, I have no idea how Discord works)

Edit: @RevoGap and I got it working. There was one idiosyncrasy, the `dac.sock` path was not correct. We had to do `find / -name dac.sock` in order to get the correct path to add to the text file. After we did that though, everything else seemed to work.

**Comment by RevoGap** [comment]
To report back: with the amazing help from morelazers, I was able to get it to work successfully on my Pod 4! I emptied it a while ago, but it primed it successfully. It is heating as we speak.
Could not have done it without morelazers' lengthy troubleshooting!! Thank you so much for this app, this should be a big help for all those who want full ownership of their Pod and its functionality. If anyone needs any help, let me know!

**Comment by throwaway31265** [comment]
@morelazers @RevoGap That's great news! What was the dac.sock path?


> Every keystroke I make in the terminal window now lights up the LED on the TC2070 for the TXD, but nothing shows up in the terminal window. After running the command and powering on the Hub 4, no message shows up saying "Hit any key to stop autoboot: 0"

What was the issue here? Did you have the minicom session opened prior to powering on the device?

**Comment by morelazers** [comment]
@throwaway31265 it was the same as the Pod 3 but with `/persistent` appended to the beginning. I don't recall the full path.

**Comment by throwaway31265** [comment]
`/persistent/deviceinfo/dac.sock` I'm guessing, I've seen it in a few different paths now. Did you do a firmware reset prior to installing?

**Comment by morelazers** [comment]
That was the path, yeah. I don't know if @RevoGap performed a firmware reset on their Pod 4. I can confirm that I did on my Pod 3 and the path was without `/persistent`.

**Comment by RevoGap** [comment]
> @morelazers @RevoGap That's great news! What was the dac.sock path?
> 
> > Every keystroke I make in the terminal window now lights up the LED on the TC2070 for the TXD, but nothing shows up in the terminal window. After running the command and powering on the Hub 4, no message shows up saying "Hit any key to stop autoboot: 0"
> 
> What was the issue here? Did you have the minicom session opened prior to powering on the device?

The pins were just connected to the wrong ones, since I used the Pod 3 board picture as a guide for the pins before I saw the Pod 4 board image. @morelazers pointed me to this and then it was easy as cake to talk to the Pod to get root

**Comment by RevoGap** [comment]
> `/persistent/deviceinfo/dac.sock` I'm guessing, I've seen it in a few different paths now. Did you do a firmware reset prior to installing?

I did not do a firmware reset, I just started from the start of the 'How to get root' guide and later switched to the installation guide from the main project GitHub page.

**Comment by throwaway31265** [comment]
@RevoGap Any chance you could share a pic of how to take off the grill on the pod 4, what the circuit board looks  like, and a pic with the tag-connect plugged in? (That way others can see an example in the docs)

**Comment by RevoGap** [comment]
I posted the pics of the board and daughter board a few posts before, which were slightly newer than nota-nota-nota's. 
Here are the pics of the pin configuration on my tag-connect, with a pic from the documentation of the tag-connect and the free-sleep GitHub project page to see which pins correspond with the pin locations on the board as described above by nota-nota-nota.
I also added pictures of the back of the mesh front with the locations and spacing of the clips, as well as how the clips connect to the frame on the front of the Pod 4 Hub unit, so others can unclip it safely. 

Had my first night's sleep on the cover: everything is working perfectly. Donated some BTC to the project: again, thank you for your hard work! Will be on the lookout for new features. Since you asked me on Reddit for requests: maybe creating the dim option for the cover light would be easy enough? Also, if it would be possible to control the heating manually with the 'button' on the sides of the cover, that would also be an amazing upgrade.

Would love it if there would be any way to get the premium subscription options to work some time in the future, for dynamic temperature control and reports, but I expect that this could be an extraordinarily if not pretty impossible feat. Any other features will be great, but the basic functionality is already way beyond what I thought was already possible. 

![IMG_2037](https://github.com/user-attachments/assets/427e81ba-6ddd-4ef8-a07b-0dd38209aeb9)
![IMG_2039](https://github.com/user-attachments/assets/a6084989-3c54-426c-94b8-33bfa24bd15b)
![IMG_2040](https://github.com/user-attachments/assets/5d2be774-19ea-49a7-9c71-09bd6ad4f942)
![IMG_2041](https://github.com/user-attachments/assets/03478be5-fc15-40d3-befe-2c7a6377382c)
![IMG_2064](https://github.com/user-attachments/assets/9c142a7c-94ae-4ad5-856d-36f5d5f1f972)
![IMG_2067](https://github.com/user-attachments/assets/f07de414-ca04-4539-a306-8dffe18fb9bd)
![IMG_2068](https://github.com/user-attachments/assets/75205909-3dbe-4b3c-bca3-a95f5233d0df)

**Comment by throwaway31265** [comment]
@RevoGap Thanks for your donation & sharing photos! 

I just deployed an update. You can set the brightness of the LED on the settings page now. 

I made some other updates as well, they're in the commit history. I'm currently looking at the sensor data for HR data and seeing some promising results validated against my apple watch.

**Comment by dallonby** [comment]
Has anyone got any ideas:

```CPU:   MediaTek MT8365
DRAM:  1 GiB
Core:  24 devices, 15 uclasses, devicetree: separate
WDT:   Started watchdog@10007000 with servicing (60s timeout)
MMC:   mmc@11230000: 0
Loading Environment from MMC... OK
In:    uart0@11002000
Out:   uart0@11002000
Err:   uart0@11002000
Net:   eth0: usb_ether
Hit any key to stop autoboot:  0
switch to partitions #0, OK
mmc0(part 0) is current device
Booting on boot_a ...
Retrieving file: /extlinux/extlinux.conf
** Unable to read file /extlinux/extlinux.conf **```

**Comment by tmandry** [comment]
**Make sure your version of minicom is updated.** I tried this using minicom 2.8 and got garbage. It was actually able to read the initial boot sequence at baud 115200, but never the uBoot / login prompt at 921600 baud. Worked perfectly with minicom 2.10, and the free-sleep instructions report 2.9 working too.
